app/src/org/commcare/tasks/ResourceRecoveryTask.java

```diff
@@ -0,0 +1,76 @@
+package org.commcare.tasks;
+
+import org.commcare.CommCareApplication;
+import org.commcare.engine.resource.ResourceInstallUtils;
+import org.commcare.resources.model.InstallCancelled;
+import org.commcare.resources.model.InstallCancelledException;
+import org.commcare.resources.model.Resource;
+import org.commcare.resources.model.ResourceTable;
+import org.commcare.resources.model.TableStateListener;
+import org.commcare.resources.model.UnreliableSourceException;
+import org.commcare.resources.model.UnresolvedResourceException;
+import org.commcare.tasks.templates.CommCareTask;
+import org.commcare.utils.AndroidCommCarePlatform;
+import org.javarosa.xml.util.UnfullfilledRequirementsException;
+
+import java.util.Vector;
+
+public abstract class ResourceRecoveryTask<Reciever>
+        extends CommCareTask<Void, Integer, Boolean, Reciever> implements TableStateListener, InstallCancelled {
+
+    public ResourceRecoveryTask(int taskId) {
+        this.taskId = taskId;
+    }
+
+    @Override
+    protected Boolean doTaskBackground(Void... voids) {
+        AndroidCommCarePlatform platform = CommCareApplication.instance().getCommCarePlatform();
+        ResourceTable global = platform.getGlobalResourceTable();
+        setTableListeners(global);
+        boolean success;
+        try {
+            success = global.recoverResources(platform, getProfileReference());
+        } catch (InstallCancelledException | UnresolvedResourceException | UnfullfilledRequirementsException e) {
+            throw new RuntimeException(e);
+        } finally {
+            unsetTableListeners(global);
+        }
+        return success;
+    }
+
+    private String getProfileReference() {
+        String profileRef = ResourceInstallUtils.getDefaultProfileRef();
+        return profileRef.split("\\?")[0];
```

Comment:
This looks a little magic-y; this is splitting the base URL and params right? Maybe a comment or more descriptive method title would be helpful?